### PR TITLE
kernel: process: reconfigure mpu on restart

### PR DESCRIPTION
**Pull request by bradjc**
If a process restarts we need to setup the MPU based on the original app break and not the one the app had when it crashed.

If an app shrank its memory before crashing then not doing this causes an immediate MPU fault if the kernel puts the stack outside of the app's old memory region.

If an app grew its memory before crashing then not doing this causes a security issue since an app can grow its heap, crash, then restart and have access to memory above its app break address.

### Testing Strategy

This pull request was tested by using the stack test 01 app on hail with the bug in libtock-c that causes the app to crash.

### TODO or Help Wanted

n/a

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: kernel/src/process.rs
Comment (hudson-ayers) [inline kernel/src/process.rs:2083]:
```suggestion
                self.original_app_break,
                self.original_kernel_memory_break,
```
(I realize this is basically equivalent, because `app_break` is set to `original_app_break` above, but this way removes an extra `get()` call and protects this mistake from being recreated if the code in this function is shuffled such that the MPU reconfig happens before changing the non-original values)

File: kernel/src/process.rs
Comment (hudson-ayers) [inline kernel/src/process.rs:2085]: This appears to still use the MPU config of the failed process, which is not guaranteed to be the same as the MPU config of a newly created process, as I understand it (`allocate_region()`, for example, modifies the config).

Should we be replacing the config with the `Default` before this?

**Comment by hudson-ayers** [comment]
I pushed a commit that I believe addresses my concerns. I believe that there are better ways to do this, but the better ways involve changing the MPU trait which goes beyond the purview of a pre-release bug fix.

Now additional allocated MPU regions for a process (such as those allocated by IPC) will not carry over past a restart of that process.

**Comment by hudson-ayers** [comment]
The QEMU "push" check only fails because this PR has not been rebased to use the new QEMU download location, but bors will not block on that check.

**Review by hudson-ayers — APPROVED** [review]
Giving my approval to brads commit, but someone should review my added commit before we merge this.

**Comment by bradjc** [comment]
Looks good to me.

After several discussions it does seem that recomputing is the approach to take.

**Pull request merged**
